cnxepub/formatters.py

```diff
@@ -426,7 +429,77 @@ def exercise_callback_factory(match, url_template,
     """Create a callback function to replace an exercise by fetching from
     a server."""
 
-    def _replace_exercises(elem):
+    def _annotate_exercise(elem, exercise, page_uuids):
+        """Annotate exercise based upon tag data"""
+        tags = exercise['items'][0].get('tags')
+        if not tags:
+            return
+
+        # Annotate exercise with required context, if it exists
+        modules, feature = [], ''
+        for tag in tags:
+            if 'context-cnxmod:' in tag:
+                modules.append(tag.split(':')[1])
+            if 'context-cnxfeature:' in tag:
+                # Note: Assuming this tag will never be duplicated with
+                # multiple values in the exercise data
+                feature = tag.split(':')[1]
+
+        # It's possible that the feature tag is not present, or present but
+        # not valid (e.g. value of "context-cnxfeature:").
+        if not feature:
+            return
+
+        # There may be multiple `context-cnxmod:{uuid}` tags, each of which
+        # could be the parent page for this exercise, a different page in this
+        # book, or even invalid altogether. We'll prefer the first, fallback
+        # to the second, and not annotate in the last case.
+
+        # Determine parent page for this exercise
+        parent_page_uuid = None
+        parent_page_elem = None
+        for parent in elem.iterancestors():
+            if parent.get('data-type') == 'page':
+                parent_page_elem = parent
+                parent_page_uuid = parent.get('id')
+                if parent_page_uuid.startswith('page_'):
+                    # Strip `page_` prefix from ID to get UUID
+                    parent_page_uuid = parent_page_uuid.split('page_')[1]
+                break
+
+        candidate_uuids = set(modules) & set(page_uuids)
+
+        # Check if the target feature ID is on the parent page for this
+        # exercise. If so, that takes priority over any context-cnxmod tag
+        # values, and should be picked even in cases where the parent page
+        # doesn't match one of the exercise tags. If the feature exists,
+        # we make sure the parent page UUID it's included in candidate_uuids
+        maybe_feature = parent_page_elem.find(
+            './/*[@id="{}"]'.format("auto_%s_%s" % (parent_page_uuid, feature))
+        )
+        if (maybe_feature is not None):
+            candidate_uuids.add(parent_page_uuid)
+
+        if len(candidate_uuids) == 0:
+            # No valid page UUIDs in exercise data
+            return
```

Comment:
I can't remember what @oscryan mentioned last week. I'm still worried about silently failing (not including the context information) when a `context-cnxfeature:` tag is set on an exercise but no valid module UUID could be found.

Reply:
Ah, right. I wasn't clear on the best assumption to make here as far as whether it's never valid or a corner case that can occur. We can `raise` here if the former is true, though.

Reply:
Per discussion with @oscryan , it is possible for these scenarios to occur (for example as content is edited and the feature ID no longer matches the content). However, instead of creating a hole for this edge case, we'll `raise` here and if it's really a valid scenario to keep the exercise w/o context, it will trigger us to remove the `cnxfeature` tag from the exercise.

Reply:
Addressed with [this change](https://github.com/openstax/cnx-epub/pull/177/commits/5cefef609abb33ced1d527d668985df398cb91ae).